[log4net] bridge + trace context injection

lachmatt opened the pull request:
## Why

Towards #3822

## What

Initial implementation of `log4net` bridge and injection of trace context.

Bytecode instrumentation for `log4net` that adds appender exporting logs in OTLP using LogsApi.

Bytecode instrumentation targets and some of the implementation similar to Datadog's `log4net` DirectLogSubmission/trace context injection implementation. @zacharycmontoya PTAL and approve from Datadog's side.

Logs bridge is **disabled** by default. In order to enable it, `OTEL_DOTNET_AUTO_LOGS_ENABLE_LOG4NET_BRIDGE` env variable needs to be set.
Trace context injection is **enabled** by default.

I added an internal doc with a short description of instrumentation added in this PR, and known limitations.

If `*Format` overloads (e.g `InfoFormat`) are used for logging, format string is used as a body of a log record, and args are added as attributes. If additionally `OTEL_DOTNET_AUTO_LOGS_INCLUDE_FORMATTED_MESSAGE` is set, `log4net.rendered_message` attribute is added with a content of a formatted message. Sample output:
```
LogRecord.Timestamp:               2024-12-11T15:03:10.8476757Z
LogRecord.TraceId:                 277283fe7ba44f2e91b0cb0b2d5f5caf
LogRecord.SpanId:                  306922dc6adef3d6
LogRecord.TraceFlags:              Recorded
LogRecord.CategoryName:            TestApplication.Log4NetBridge.Program
LogRecord.Severity:                Info
LogRecord.SeverityText:            INFO
LogRecord.Body:                    {0}, {1} at {2:t}!
LogRecord.Attributes (Key:Value):
    log4net.rendered_message: Hello, world at 16:03!
    0: Hello
    1: world
    2: 12/11/2024 16:03:10
```

Notes:
Most of the LogsApi from OpenTelemetry .NET SDK [is not public](https://github.com/open-telemetry/opentelemetry-dotnet/blob/f9a0b4c657e24cc8c52a54795e13801eb5548c22/docs/diagnostics/experimental-apis/OTEL1001.md) in non-rc builds. Due to this, and to achieve acceptable peformance, expressions are used to initialize delegates used for sending logs.

In order to avoid adding multiple appenders serving the same purpose, as a simple workaround for now, this integration is enabled only if `ILogger` integration is not initialized.

## Tests

Included in PR.

## Checklist

<!-- All items should be verified and marked as done.
     ~~strikethrough~~ if an item doesn't apply to the introduced changes. -->

- [x] `CHANGELOG.md` is updated.
- [x] Documentation is updated.
- [x] New features are covered by tests.

Kielek commented on src/OpenTelemetry.AutoInstrumentation/Configurations/LogInstrumentation.cs:
```diff
@@ -11,5 +11,10 @@ internal enum LogInstrumentation
     /// ILogger instrumentation.
     /// </summary>
-    ILogger
+    ILogger,
+
+    /// <summary>
+    /// Log4Net instrumentation.
+    /// </summary>
+    Log4Net,
```
To make it consistent with other signals

```suggestion
    ILogger = 0,

    /// <summary>
    /// Log4Net instrumentation.
    /// </summary>
    Log4Net = 1,
```

Kielek commented on src/OpenTelemetry.AutoInstrumentation/Instrumentation.cs:
```diff
@@ -166,6 +172,22 @@ public static void Initialize()
+                var createLoggerProviderBuilderMethod = typeof(Sdk).GetMethod("CreateLoggerProviderBuilder", BindingFlags.Static | BindingFlags.NonPublic)!;
+                var loggerProviderBuilder = createLoggerProviderBuilderMethod.Invoke(null, null) as LoggerProviderBuilder;
+
+                // TODO: plugins support
```
:)

Kielek commented on src/OpenTelemetry.AutoInstrumentation/Instrumentations/Log4Net/Integrations/AppenderCollectionIntegration.cs:
```diff
@@ -0,0 +1,33 @@
+methodName: "ToArray",
+returnTypeName: "log4net.Appender.IAppender[]",
+parameterTypeNames: new string[0],
+minimumVersion: "2.0.0",
+maximumVersion: "3.*.*",
```
Just double check, there is no changes between 2.0.0 and 3.0.0 in instrumented method?

lachmatt replied:
That's correct.

Kielek commented on test/test-applications/integrations/TestApplication.Log4NetBridge/Program.cs:
```diff
@@ -0,0 +1,72 @@
+            else if (logApiName == "ILogger")
+            {
+                LogUsingILogger();
+            }
```
Add NotSupportedException or something like that when logapiname != log4net and ilogger.

lachmatt commented:
I'm still looking into one problematic scenario, and extracting message template and args that were used for logging, but any feedback at this point would be welcome.

Kielek commented on .cspell/company-terms.txt:
```diff
@@ -7,3 +7,4 @@ triager
 Zipkin
 parentbased
 traceidratio
+appender
```
alphabetical order?

Kielek commented on test/OpenTelemetry.AutoInstrumentation.Tests/OpenTelemetry.AutoInstrumentation.Tests.csproj:
```diff
@@ -3,6 +3,7 @@
   <ItemGroup>
     <PackageReference Include="Newtonsoft.Json" />
     <PackageReference Include="Xunit.SkippableFact" />
+    <PackageReference Include="log4net" />
```
I would consider moving tests related to the dependencies to the separate package. To avoid brining here transitive dependencies.

Kielek commented on test/IntegrationTests/Log4NetBridgeTests.cs:
```diff
@@ -0,0 +1,146 @@
+        collector.AssertExpectations();
+    }
+
+#if !NETFRAMEWORK
```
```suggestion
#if NET
```

Kielek commented on src/OpenTelemetry.AutoInstrumentation/Instrumentations/Log4Net/Integrations/AppenderCollectionIntegration.cs:
```diff
@@ -0,0 +1,42 @@
+// SPDX-License-Identifier: Apache-2.0
+
+using OpenTelemetry.AutoInstrumentation.CallTarget;
+#if !NETFRAMEWORK
```
```suggestion
#if NET
```

Kielek commented on src/OpenTelemetry.AutoInstrumentation/Instrumentations/Log4Net/Integrations/AppenderCollectionIntegration.cs:
```diff
@@ -0,0 +1,42 @@
+    internal static CallTargetReturn<TReturn> OnMethodEnd<TTarget, TReturn>(TTarget instance, TReturn returnValue, Exception exception, in CallTargetState state)
+    {
+        if (
+            #if !NETFRAMEWORK
```
```suggestion
#if NET
```

zacharycmontoya commented on src/OpenTelemetry.AutoInstrumentation/Instrumentations/Log4Net/OpenTelemetryLog4NetAppender.cs:
```diff
@@ -0,0 +1,126 @@
+    internal static int MapLogLevel(int levelValue)
+#pragma warning restore SA1202
+    {
+        return levelValue switch
```
Is there a source for this mapping? I was thinking we could follow the [example of Java instrumentation](https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/57c7cf2ad5f7272c1eb83b9816e652bf832c91d4/instrumentation/log4j/log4j-appender-2.17/library/src/main/java/io/opentelemetry/instrumentation/log4j/appender/v2_17/internal/LogEventMapper.java#L226) to call `GetStandardLevel` on the LogLevel (which returns a general DEBUG/INFO/WARNING/ERROR/FATAL) and return the general category instead of specific subcategories

zacharycmontoya replied:
Also, can we introduce an enum to map to the OpenTelemetry API Log levels and use those for the return values?

lachmatt replied:
I used a guidance from [the specification repository](https://github.com/open-telemetry/opentelemetry-specification/blob/a265ae0628177be25dc477ea8fe200f1c825b871/specification/logs/data-model.md?plain=1#L315-L323) for the mapping.
At the same time I'm fine with changing mapping to follow the approach outlined above.

lachmatt replied:
Mapping changed to return general categories.

zacharycmontoya commented on test/IntegrationTests/Log4NetBridgeTests.cs:
```diff
@@ -0,0 +1,146 @@
+    [Theory]
+    [Trait("Category", "EndToEnd")]
+    [MemberData(nameof(LibraryVersion.log4net), MemberType = typeof(LibraryVersion))]
+    public async Task SubmitLogs_ThroughILoggerBridge_WhenLog4NetIsUsedAsILoggerAppenderForLogging_WithoutDuplicates(string packageVersion)
```
Is this testing anything different than the above ILogger one?

lachmatt replied:
Test above won't fail if there are additional log records, which were not expected.

pellared commented on docs/config.md:
```diff
@@ -200,6 +200,7 @@ due to lack of stable semantic convention.
 | ID        | Instrumented library                                                                                                            | Supported versions | Instrumentation type   | Status                                                                                                                            |
 |-----------|---------------------------------------------------------------------------------------------------------------------------------|--------------------|------------------------|-----------------------------------------------------------------------------------------------------------------------------------|
 | `ILOGGER` | [Microsoft.Extensions.Logging](https://www.nuget.org/packages/Microsoft.Extensions.Logging) **Not supported on .NET Framework** | ≥8.0.0             | bytecode or source [1] | [Experimental](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/versioning-and-stability.md) |
+| `LOG4NET` | [log4net](https://www.nuget.org/packages/log4net)                                                                               | ≥2.0.0 && < 4.0.0  | bytecode               | [Experimental](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/versioning-and-stability.md) |
```
I think that the mapping (translation form log4net to OTel Logs) should be documented.

lachmatt replied:
Added in https://github.com/open-telemetry/opentelemetry-dotnet-instrumentation/pull/3825/commits/735deacf9608861c6d797b80cb1851f45d9d1be3

pellared commented on docs/config.md:
```diff
@@ -200,6 +200,7 @@ due to lack of stable semantic convention.
 | ID        | Instrumented library                                                                                                            | Supported versions | Instrumentation type   | Status                                                                                                                            |
 |-----------|---------------------------------------------------------------------------------------------------------------------------------|--------------------|------------------------|-----------------------------------------------------------------------------------------------------------------------------------|
 | `ILOGGER` | [Microsoft.Extensions.Logging](https://www.nuget.org/packages/Microsoft.Extensions.Logging) **Not supported on .NET Framework** | ≥8.0.0             | bytecode or source [1] | [Experimental](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/versioning-and-stability.md) |
+| `LOG4NET` | [log4net](https://www.nuget.org/packages/log4net)                                                                               | ≥2.0.0 && < 4.0.0  | bytecode               | [Experimental](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/versioning-and-stability.md) |
```
Side note: The section should be "Logs bridges" or "Logs appenders" and not "Logs instrumentations". The "Logs instrumentations" may come in future. E.g. OTel may add Logs instrumentations for HTTP Client.

lachmatt commented:
I added trace context injection, extraction of format string and args. I updated the PR description to better reflect scope of  the changes in this PR.

zacharycmontoya commented on docs/internal/log4net.md:
```diff
@@ -0,0 +1,60 @@
+
+> [!IMPORTANT]
+> log4net bridge and trace context injection are experimental features.
+> Both instrumentations can be disabled by setting `OTEL_DOTNET_AUTO_LOGS_LOG4NET_INSTRUMENTATION_ENABLED` to `false`.
```
I don't see the env variable `OTEL_DOTNET_AUTO_LOGS_LOG4NET_INSTRUMENTATION_ENABLED` anywhere, only `OTEL_DOTNET_AUTO_LOGS_ENABLE_LOG4NET_BRIDGE`. Am I missing something?

Kielek replied:
We have two levels
1. Injecting context to the log files by standard funtionality https://github.com/open-telemetry/opentelemetry-dotnet-instrumentation/pull/3825/files#diff-77b3519370ff48f6caf7c316f0ece966f2f0871c40e9381d451d2c5b1c76b40fR124 
2. Bridging everything to the exporter (typically OTLP) by `OTEL_DOTNET_AUTO_LOGS_ENABLE_LOG4NET_BRIDGE`

Kielek replied:
`OTEL_DOTNET_AUTO_LOGS_LOG4NET_INSTRUMENTATION_ENABLED` is part of our standard functionality

zacharycmontoya reviewed (approved):
Overall LGTM, good job putting together the changes and the testing
